### ISSUE TITLE
[Security] Rename provider key

### ISSUE THREAD

**Issue by wouterj** [symfony/symfony#15207]
As also explained in https://github.com/symfony/symfony/pull/15141 , provider key is a very confusing term in the Security component. A lot of people (including me a week ago) don't know what this property does and what value it should have.

I think renaming this property and adding better PHPdocs can improve things a lot on this subject. However, finding a good name is not easy. This is were I need feedback (and other suggestions).

**In short**: The provider key is used to group a set of authentication providers. In the full stack framework, this means you can have multiple firewall (each firewall having its own group of authentication providers). Being authenticated for one firewall doesn't mean you are authenticated for the other.

**My suggestions**: context (consistent with the name of the firewall config); provider group.

**Comment by iltar** [comment]
I would like to add a small note of why I agree with 'context'.

In the firewall configuration, the name of the firewall will be the 'context' as mentioned here. However, you're able to provide a 'context' option per firewall which allows you to group firewalls under the same 'context' so tokens can be shared.

http://symfony.com/doc/current/reference/configuration/security.html#firewall-context

**Comment by javiereguiluz** [comment]
@WouterJ I appreciate your efforts to improve both the code and the documentation of the security component. However, I don't know if this is the best strategy. You are proposing a lot of independent and small changes (like the rename done in #15141). I'd prefer if we could think of a "big picture" of the new security component and layout all the needed changes.

I'd like to have three different ways of using the component:

  * A "fast track" for all the common features: login form, in-memory users, encoding passwords, login users programmatically, login with third-party sites (Twitter, Facebook, etc.), load users stored in a database, etc.
  * A "clear path" for uncommon but not rare features: X509 certificates, two-factor authentication, etc.
  * A "do-it-yourself path" for very specific or rare needs.

My guess is that we can improve the "fast track" a lot. For instance, if your users are stored in a database, login them programmatically should be as easy as `$this->get('security')->login($userObject, $firewallName = 'main')` but now we have to deal with `UsernamePasswordToken` and other low-level details people don't care about.

**Comment by WouterJ** [comment]
@javiereguiluz there are already very big PRs fixing the "fast track". For instance, https://github.com/symfony/symfony/pull/13062 (direct authentication) and https://github.com/symfony/symfony/pull/14673 (guard).

These PRs are there to fix just the minor things that are causing headages when using the Security component. I've seen questions about the provider key often enough on IRC to know that there is room for improvement there.

---

The problem with `context`, is that it is a very general term again. I'm not sure how much a parameter called `context` is going to improve over a parameter called `providerKey`.

**Comment by javiereguiluz** [comment]
I'll try to explain it again with an example.

In this PR you propose to rename this thing to `context` and that certainly seems the right thing to do. But if you create a meta-issue listing the 15 small changes we plan to introduce in the security component, we'll have the whole perspective to decide if `context` is the best rename or not.

**Comment by fabpot** [comment]
I don't see how `context` would be better than `providerKey`. `context` is one of those overloaded keyword that does not convey any special meaning.

**Comment by iltar** [comment]
@fabpot going by that logic, would it be an interesting idea to rename the configuration option something else, together with the variable?

**Comment by WouterJ** [comment]
@fabpot that's also why I like "providerGroup" more. I think it holds more meaning than "providerKey" (personally, I was thinking about one provider when talking about "key", while "group" kind of makes you think about multiple providers).

**Comment by WouterJ** [comment]
ping @symony/deciders (useless mention, as it doesn't work, but anyway...) can we make a decision on this please? I would like to fix this name that's causing confusion before 2.8/3.0

**Comment by stof** [comment]
@WouterJ it could work if the org name was right :smile: (it depends whether the org opted in the new team management)

**Comment by fabpot** [comment]
Now that Guard is in, I don't think we need to rename `providerKey`. `providerGroup` does not look better to me either. Just my 2c.

**Comment by fabpot** [comment]
A few years later, what do you think we should do ?

**Comment by wouterj** [comment]
This issue is more relevant than ever.... and I secretly already did this in my first big PR :roll_eyes: . In all code related to the new system, `$providerKey` has been renamed to `$firewallName`. As authenticators live in Http, the concept of firewall can be used throughout the system. Also, "providers" no longer exists.

We should probably check if we use it consistently in all code that will remain in the Security package in Symfony 6.0 (e.g. I don't think we should rename parameters used by the legacy authentication providers).

**Comment by fabpot** [comment]
Actually, I saw the rename and `firewallName` is a much better name. My comment was more about closing it here as we will switch to the new system where the issue does not exist anymore. Not sure it's worth renaming the old system. WDYT?

**Comment by wouterj** [comment]
Yeah, I don't think it's worth updating old code that will be removed in 6.0 (i.e. the old manager and all authentication -providers and -listeners).

From a quick grep (And filtering out the old system classes), these files still use `providerKey` somewhere:
```
src/Symfony/Bundle/SecurityBundle/Tests/DependencyInjection/SecurityExtensionTest.php
src/Symfony/Component/Security/Core/Authentication/Token/PreAuthenticatedToken.php
src/Symfony/Component/Security/Core/Authentication/Token/RememberMeToken.php
src/Symfony/Component/Security/Core/Authentication/Token/SwitchUserToken.php
src/Symfony/Component/Security/Core/Authentication/Token/UsernamePasswordToken.php
src/Symfony/Component/Security/Http/Authentication/CustomAuthenticationSuccessHandler.php
src/Symfony/Component/Security/Http/Authentication/DefaultAuthenticationSuccessHandler.php
src/Symfony/Component/Security/Http/Event/LoginSuccessEvent.php
src/Symfony/Component/Security/Http/Firewall/ExceptionListener.php
src/Symfony/Component/Security/Http/Firewall/SwitchUserListener.php
src/Symfony/Component/Security/Http/RememberMe/AbstractRememberMeServices.php
src/Symfony/Component/Security/Http/Tests/EventListener/RememberMeListenerTest.php
src/Symfony/Component/Security/Http/Tests/EventListener/SessionStrategyListenerTest.php
src/Symfony/Component/Security/Http/Tests/Firewall/SwitchUserListenerTest.php
src/Symfony/Component/Security/Http/Tests/Util/TargetPathTraitTest.php
src/Symfony/Component/Security/Http/Util/TargetPathTrait.php
```

Maybe we can keep this issue open till we've checked them (and renamed them). I'm not sure about the Symfony policy on named arguments in PHP 8, but it's probably worth doing this before the 5.2 release? (or even consider this a bugfix and do it in 4.4?)
